app/controllers/bulkrax/importers_controller.rb

```diff
@@ -128,7 +138,41 @@ def set_importer
 
       # Only allow a trusted parameter "white list" through.
       def importer_params
-        params.require(:importer).permit(:name, :admin_set_id, :user_id, :frequency, :parser_klass, :limit, field_mapping: {}, parser_fields: {})
+        params.require(:importer).permit(:name, :admin_set_id, :user_id, :frequency, :parser_klass, :limit, :selected_files, field_mapping: {}, parser_fields: {})
+      end
+
+      # @todo - investigate getting directory structure
+      # @todo - investigate using perform_later, and haveing the importer check for
+      #   DownloadCloudFileJob before it starts
+      def retrieve_files(files)
```

Comment:
it feels like retrieve_files should move to a method on the parser, which would give you a simpler way to handle the different importer needs.

having said that, the files.each_pair code would work just as well for the case of 1 as it does for the case of many and could be simplified to be the same in both cases, leaving you with only the file path logic as different.

Reply:
I've refactored and tried to make the code a little clearer.